File: agents/node/src/sanitize.ts
// DB statement sanitization, a port of the Go agent's openlogsql.Sanitize (agents/go/openlogsql/sanitize.go)
// so both agents send the same db.query.text for the same statement (apm.md §7 normalizes again on the backend).

/** Upper bound of db.query.text (bytes of UTF-16 code units; statements are ASCII in practice), like the Go agent. */
export const MAX_QUERY_TEXT = 4096;

const isDigit = (c: number): boolean => c >= 48 && c <= 57;
const isHex = (c: number): boolean => isDigit(c) || (c >= 97 && c <= 102) || (c >= 65 && c <= 70);
const isIdentStart = (c: number): boolean => c === 95 || (c >= 97 && c <= 122) || (c >= 65 && c <= 90);
const isIdent = (c: number): boolean => isIdentStart(c) || isDigit(c) || c === 36;
const isNonASCII = (c: number): boolean => c >= 128;

function identBefore(q: string, i: number): boolean {
  if (i === 0) return false;
  const p = q.charCodeAt(i - 1);
  return isIdent(p) || isNonASCII(p);
}

/**
 * Normalizes a SQL statement so it carries no literal values and statements differing only in values group
 * together: string literals ('…', E'…', and "…" for MySQL) → ?; numeric and hex literals → ? (identifiers such as
 * t1 or col_2 are kept); IN (?, ?) and VALUES (?), (?) lists collapse to (?); comments are removed and whitespace
 * is collapsed. Placeholders ($1, ?, :name, @p1) are kept. `system` is the db.system.name value; it only changes
 * how double-quoted text (identifier except for mysql) and `#` comments (mysql) are treated.
 */
export function sanitizeSQL(q: string, system = ''): string {
  const mysql = system === 'mysql' || system === 'mariadb';
  const out: string[] = [];
  let len = 0;
  let space = false;
  const emit = (s: string): void => {
    if (space && len > 0) {
      out.push(' ');
      len++;
    }
    space = false;
    out.push(s);
    len += s.length;
  };
  const n = q.length;
  let i = 0;
  while (i < n) {
    const c = q.charCodeAt(i);
    const next = i + 1 < n ? q.charCodeAt(i + 1) : -1;
    if (c === 32 || c === 9 || c === 10 || c === 13 || c === 12) {
      space = true;
      i++;
    } else if (c === 45 /* - */ && next === 45) {
      while (i < n && q.charCodeAt(i) !== 10) i++;
      space = true;
    } else if (c === 35 /* # */ && mysql) {
      while (i < n && q.charCodeAt(i) !== 10) i++;
      space = true;
    } else if (c === 47 /* / */ && next === 42 /* * */) {
      const j = q.indexOf('*/', i + 2);
      i = j < 0 ? n : j + 2;
      space = true;
    } else if (
      c === 39 ||
      (c === 34 && mysql) ||
      ('EeNnXxBb'.includes(q[i]) && next === 39 && !identBefore(q, i))
    ) {
      const prefix = c !== 39 && c !== 34 ? q[i] : '';
      if (prefix) i++;
      const quote = q.charCodeAt(i);
      i++;
      while (i < n) {
        const d = q.charCodeAt(i);
        if (d === 92 /* \ */ && i + 1 < n && quote === 39 && (prefix === 'E' || prefix === 'e' || mysql)) {
          i += 2;
          continue;
        }
        if (d === quote) {
          if (i + 1 < n && q.charCodeAt(i + 1) === quote) {
            i += 2;
            continue;
          }
          i++;
          break;
        }
        i++;
      }
      emit('?');
    } else if (c === 36 /* $ */ && next === 36) {
      const j = q.indexOf('$$', i + 2);
      i = j < 0 ? n : j + 2;
      emit('?');
    } else if (c === 34 || c === 96 || c === 91) {
      const closer = c === 91 ? ']' : q[i];
      const j = q.indexOf(closer, i + 1);
      if (j < 0) {
        emit(q.slice(i));
        i = n;
      } else {
        emit(q.slice(i, j + 1));
        i = j + 1;
      }
    } else if (isDigit(c) || (c === 46 && next >= 0 && isDigit(next))) {
      if (identBefore(q, i)) {
        emit(q[i]);
        i++;
        continue;
      }
      if (c === 48 && (next === 120 || next === 88)) {
        i += 2;
        while (i < n && isHex(q.charCodeAt(i))) i++;
      } else {
        while (i < n && (isDigit(q.charCodeAt(i)) || q.charCodeAt(i) === 46)) i++;
        if (i < n && (q[i] === 'e' || q[i] === 'E')) {
          let k = i + 1;
          if (k < n && (q[k] === '+' || q[k] === '-')) k++;
          if (k < n && isDigit(q.charCodeAt(k))) {
            i = k;
            while (i < n && isDigit(q.charCodeAt(i))) i++;
          }
        }
      }
      emit('?');
    } else if (
      (c === 36 || c === 64 || c === 58) &&
      next >= 0 &&
      (isDigit(next) || isIdentStart(next)) &&
      !(c === 58 && i > 0 && q.charCodeAt(i - 1) === 58)
    ) {
      let j = i + 1;
      while (j < n && isIdent(q.charCodeAt(j))) j++;
      emit(q.slice(i, j));
      i = j;
    } else if (isIdentStart(c) || isNonASCII(c)) {
      let j = i + 1;
      while (j < n && (isIdent(q.charCodeAt(j)) || isNonASCII(q.charCodeAt(j)))) j++;
      emit(q.slice(i, j));
      i = j;
    } else {
      if (c === 44 || c === 41) space = false;
      emit(q[i]);
      if (c === 40) space = false;
      i++;
    }
  }
  return collapseLists(out.join(''));
}

/** "(?, ?, ?)" → "(?)" and "VALUES (?), (?)" → "VALUES (?)". */
function collapseLists(s: string): string {
  if (!s.includes('?,')) return s;
  let b = '';
  let i = 0;
  while (i < s.length) {
    if (s[i] === '(') {
      let j = i + 1;
      let ok = false;
      while (j < s.length) {
        if (s[j] === '?') {
          j++;
          ok = true;
          if (j < s.length && s[j] === ',') {
            j++;
            if (j < s.length && s[j] === ' ') j++;
            continue;
          }
        }
        break;
      }
      if (ok && j < s.length && s[j] === ')') {
        b += '(?)';
        i = j + 1;
        for (;;) {
          let k = i;
          if (k < s.length && s[k] === ',') {
            k++;
            if (k < s.length && s[k] === ' ') k++;
            if (k < s.length && s[k] === '(') {
              let m = k + 1;
              while (m < s.length && (s[m] === '?' || s[m] === ',' || s[m] === ' ')) m++;
              if (m < s.length && s[m] === ')' && m > k + 1) {
                i = m + 1;
                continue;
              }
            }
          }
          break;
        }
        continue;
      }
    }
    b += s[i];
    i++;
  }
  return b;
}

/**
 * Key/value store statement (redis, memcached): the command word upper-cased and every argument replaced by `?`
 * (apm.md §7: `GET products:42` → `GET ?`). Long argument lists keep at most 32 placeholders plus `…`.
 */
export function sanitizeKeyValue(command: string, args: ReadonlyArray<unknown> = []): string {
  const parts = [String(command).toUpperCase()];
  const shown = Math.min(args.length, 32);
  for (let k = 0; k < shown; k++) parts.push('?');
  if (args.length > shown) parts.push('…');
  return parts.join(' ');
}

/** Truncates to MAX_QUERY_TEXT without splitting a surrogate pair. */
export function truncateQueryText(s: string, max = MAX_QUERY_TEXT): string {
  if (s.length <= max) return s;
  let end = max;
  const code = s.charCodeAt(end - 1);
  if (code >= 0xd800 && code <= 0xdbff) end--;
  return s.slice(0, end);
}

File: agents/python/src/openlog_agent/sanitize.py
"""DB statement sanitization.

A port of the Go agent's ``openlogsql.Sanitize`` (agents/go/openlogsql/sanitize.go) and the Node.js agent's
``sanitizeSQL`` so every openlog agent sends the same ``db.query.text`` for the same statement (apm.md §7 normalizes
again on the backend). Shared test cases: tests/unit/test_sanitize.py.
"""

from __future__ import annotations

from typing import Sequence

#: Upper bound of db.query.text in UTF-8 bytes, like the Go agent.
MAX_QUERY_TEXT_BYTES = 4096

_SPACE = frozenset(" \t\n\r\f")
_PREFIXES = frozenset("EeNnXxBb")


def _is_digit(c: str) -> bool:
    return "0" <= c <= "9"


def _is_hex(c: str) -> bool:
    return _is_digit(c) or "a" <= c <= "f" or "A" <= c <= "F"


def _is_ident_start(c: str) -> bool:
    return c == "_" or "a" <= c <= "z" or "A" <= c <= "Z"


def _is_ident(c: str) -> bool:
    return _is_ident_start(c) or _is_digit(c) or c == "$"


def _is_non_ascii(c: str) -> bool:
    return ord(c) >= 128


def _ident_before(q: str, i: int) -> bool:
    if i == 0:
        return False
    p = q[i - 1]
    return _is_ident(p) or _is_non_ascii(p)


def sanitize_sql(q: str, system: str = "") -> str:
    """Normalizes a SQL statement so it carries no literal values.

    String literals ('…', E'…', and "…" for MySQL) become ``?``; numeric and hex literals become ``?`` (identifiers
    such as ``t1`` or ``col_2`` are kept); ``IN (?, ?)`` and ``VALUES (?), (?)`` lists collapse to ``(?)``; comments
    are removed and whitespace is collapsed. Placeholders (``$1``, ``?``, ``:name``, ``@p1``, ``%s``) are kept.
    ``system`` is the db.system.name value; it only changes how double-quoted text (identifier except for mysql)
    and ``#`` comments (mysql) are treated.
    """
    mysql = system in ("mysql", "mariadb")
    out: list[str] = []
    length = 0
    space = False
    n = len(q)
    i = 0

    def emit(s: str) -> None:
        nonlocal space, length
        if space and length > 0:
            out.append(" ")
            length += 1
        space = False
        out.append(s)
        length += len(s)

    while i < n:
        c = q[i]
        nxt = q[i + 1] if i + 1 < n else ""
        if c in _SPACE:
            space = True
            i += 1
        elif c == "-" and nxt == "-":
            j = q.find("\n", i)
            i = n if j < 0 else j
            space = True
        elif c == "#" and mysql:
            j = q.find("\n", i)
            i = n if j < 0 else j
            space = True
        elif c == "/" and nxt == "*":
            j = q.find("*/", i + 2)
            i = n if j < 0 else j + 2
            space = True
        elif c == "'" or (c == '"' and mysql) or (c in _PREFIXES and nxt == "'" and not _ident_before(q, i)):
            prefix = c if c not in ("'", '"') else ""
            if prefix:
                i += 1
            quote = q[i]
            escapes = quote == "'" and (prefix in ("E", "e") or mysql)
            i += 1
            while i < n:
                d = q[i]
                if d == "\\" and i + 1 < n and escapes:
                    i += 2
                    continue
                if d == quote:
                    if i + 1 < n and q[i + 1] == quote:
                        i += 2
                        continue
                    i += 1
                    break
                i += 1
            emit("?")
        elif c == "$" and nxt == "$":
            j = q.find("$$", i + 2)
            i = n if j < 0 else j + 2
            emit("?")
        elif c in ('"', "`", "["):
            closer = "]" if c == "[" else c
            j = q.find(closer, i + 1)
            if j < 0:
                emit(q[i:])
                i = n
            else:
                emit(q[i : j + 1])
                i = j + 1
        elif _is_digit(c) or (c == "." and nxt != "" and _is_digit(nxt)):
            if _ident_before(q, i):
                emit(c)
                i += 1
                continue
            if c == "0" and nxt in ("x", "X"):
                i += 2
                while i < n and _is_hex(q[i]):
                    i += 1
            else:
                while i < n and (_is_digit(q[i]) or q[i] == "."):
                    i += 1
                if i < n and q[i] in ("e", "E"):
                    k = i + 1
                    if k < n and q[k] in ("+", "-"):
                        k += 1
                    if k < n and _is_digit(q[k]):
                        i = k
                        while i < n and _is_digit(q[i]):
                            i += 1
            emit("?")
        elif (
            c in ("$", "@", ":")
            and nxt != ""
            and (_is_digit(nxt) or _is_ident_start(nxt))
            and not (c == ":" and i > 0 and q[i - 1] == ":")
        ):
            j = i + 1
            while j < n and _is_ident(q[j]):
                j += 1
            emit(q[i:j])
            i = j
        elif _is_ident_start(c) or _is_non_ascii(c):
            j = i + 1
            while j < n and (_is_ident(q[j]) or _is_non_ascii(q[j])):
                j += 1
            emit(q[i:j])
            i = j
        else:
            if c in (",", ")"):
                space = False
            emit(c)
            if c == "(":
                space = False
            i += 1
    return _collapse_lists("".join(out))


def _collapse_lists(s: str) -> str:
    """``(?, ?, ?)`` → ``(?)`` and ``VALUES (?), (?)`` → ``VALUES (?)``."""
    if "?," not in s:
        return s
    b: list[str] = []
    n = len(s)
    i = 0
    while i < n:
        if s[i] == "(":
            j = i + 1
            ok = False
            while j < n:
                if s[j] == "?":
                    j += 1
                    ok = True
                    if j < n and s[j] == ",":
                        j += 1
                        if j < n and s[j] == " ":
                            j += 1
                        continue
                break
            if ok and j < n and s[j] == ")":
                b.append("(?)")
                i = j + 1
                while True:
                    k = i
                    if k < n and s[k] == ",":
                        k += 1
                        if k < n and s[k] == " ":
                            k += 1
                        if k < n and s[k] == "(":
                            m = k + 1
                            while m < n and s[m] in "?, ":
                                m += 1
                            if m < n and s[m] == ")" and m > k + 1:
                                i = m + 1
                                continue
                    break
                continue
        b.append(s[i])
        i += 1
    return "".join(b)


def sanitize_key_value(command: str, args: Sequence[object] = ()) -> str:
    """Key/value store statement (redis, memcached): ``GET products:42`` → ``GET ?``.

    The command word is upper-cased and every argument replaced by ``?``; long argument lists keep at most 32
    placeholders plus ``…`` (apm.md §7).
    """
    parts = [str(command).upper()]
    shown = min(len(args), 32)
    parts.extend("?" * shown)
    if len(args) > shown:
        parts.append("…")
    return " ".join(parts)


def truncate_query_text(s: str, max_bytes: int = MAX_QUERY_TEXT_BYTES) -> str:
    """Truncates to ``max_bytes`` UTF-8 bytes without splitting a character (like the Go agent)."""
    if len(s) * 4 <= max_bytes:
        return s
    b = s.encode("utf-8", "surrogatepass")
    if len(b) <= max_bytes:
        return s
    return b[:max_bytes].decode("utf-8", "ignore")
